src/probability/kernel/composition.lean

```diff
@@ -0,0 +1,721 @@
+  { have : function.uncurry (λ a b, η (a, b) {c : γ | (b, c) ∈ s})
+      = λ p, η p {c : γ | (p.2, c) ∈ s},
+    { ext1 p,
+      have hp_eq_mk : p = (p.fst, p.snd) := prod.mk.eta.symm,
+      rw [hp_eq_mk, function.uncurry_apply_pair], },
+    rw this,
+    exact measurable_prod_mk_mem η (measurable_fst.snd.prod_mk measurable_snd hs), },
+  exact measurable_lintegral κ h_meas,
+end
+
+lemma measurable_comp_prod_fun (κ : kernel α β) [is_s_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_s_finite_kernel η] (hs : measurable_set s) :
+  measurable (λ a, comp_prod_fun κ η a s) :=
+begin
+  simp_rw comp_prod_fun_tsum_right κ η _ hs,
+  refine measurable.ennreal_tsum (λ n, _),
+  simp only [comp_prod_fun],
+  have h_meas : measurable (function.uncurry (λ a b, seq η n (a, b) {c : γ | (b, c) ∈ s})),
+  { have : function.uncurry (λ a b, seq η n (a, b) {c : γ | (b, c) ∈ s})
+      = λ p, seq η n p {c : γ | (p.2, c) ∈ s},
+    { ext1 p,
+      have hp_eq_mk : p = (p.fst, p.snd) := prod.mk.eta.symm,
+      rw [hp_eq_mk, function.uncurry_apply_pair], },
+    rw this,
+    exact measurable_prod_mk_mem (seq η n) (measurable_fst.snd.prod_mk measurable_snd hs), },
+  exact measurable_lintegral κ h_meas,
+end
+
+/-- Composition-Product of kernels. It verifies
+`∫⁻ bc, f bc ∂(comp_prod κ η a) = ∫⁻ b, ∫⁻ c, f (b, c) ∂(η (a, b)) ∂(κ a)`
+(see `lintegral_comp_prod`). -/
+noncomputable
+def comp_prod (κ : kernel α β) [is_s_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_s_finite_kernel η] :
+  kernel α (β × γ) :=
+{ val := λ a, measure.of_measurable (λ s hs, comp_prod_fun κ η a s) (comp_prod_fun_empty κ η a)
+    (comp_prod_fun_Union κ η a),
+  property :=
+  begin
+    refine measure.measurable_of_measurable_coe _ (λ s hs, _),
+    have : (λ a, measure.of_measurable (λ s hs, comp_prod_fun κ η a s) (comp_prod_fun_empty κ η a)
+        (comp_prod_fun_Union κ η a) s) = λ a, comp_prod_fun κ η a s,
+    { ext1 a, rwa measure.of_measurable_apply, },
+    rw this,
+    exact measurable_comp_prod_fun κ η hs,
+  end, }
+
+localized "infix (name := kernel.comp_prod) ` ⊗ₖ `:100 := probability_theory.kernel.comp_prod" in
+  probability_theory
+
+lemma comp_prod_apply_eq_comp_prod_fun (κ : kernel α β) [is_s_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_s_finite_kernel η] (a : α) (hs : measurable_set s) :
+  (κ ⊗ₖ η) a s = comp_prod_fun κ η a s :=
+begin
+  rw [comp_prod],
+  change measure.of_measurable (λ s hs, comp_prod_fun κ η a s) (comp_prod_fun_empty κ η a)
+    (comp_prod_fun_Union κ η a) s = ∫⁻ b, η (a, b) {c | (b, c) ∈ s} ∂(κ a),
+  rw measure.of_measurable_apply _ hs,
+  refl,
+end
+
+lemma comp_prod_apply (κ : kernel α β) [is_s_finite_kernel κ] (η : kernel (α × β) γ)
+  [is_s_finite_kernel η] (a : α) (hs : measurable_set s) :
+  (κ ⊗ₖ η) a s = ∫⁻ b, η (a, b) {c | (b, c) ∈ s} ∂(κ a) :=
+comp_prod_apply_eq_comp_prod_fun κ η a hs
+
+/-- Integral against the composition-product of two kernels. -/
```

Comment:
```suggestion
/-- Lebesgue integral against the composition-product of two kernels. -/
```
(It would also make sense to have Bochner integral versions of these, but definitely in another PR!)